### PR TITLE
Support Nemotron DP attention and MTP

### PR DESCRIPTION
**Pull request by Zhichenzzz**
## Motivation

Enable **data-parallel (DP) attention** for Nemotron-H. Nemotron-H is a hybrid model (Mamba2 + full attention + MoE), and DP attention previously only supported standard attention + MLP transformers. So Nemotron-H ran attention and Mamba in **global TP** with no path for the DP layout — there was no gather/scatter between the DP-local (attn-TP) layout and the full-TP MoE layout, and the per-rank token padding that DP attention adds for collective alignment was not handled in the Mamba recurrent path (padding rows would write KV cache, update Mamba states, and corrupt residuals).

This PR makes attention **and** Mamba run data-parallel across attention-TP subgroups (cutting KV / Mamba-state duplication) while MoE/MLP stay in full TP.

## Modifications

- Run `MambaMixer2` and attention on the **attention-TP group** (`attn_tp_rank/size`) instead of the global TP group when DP attention is enabled; the gated RMSNorm variance reduction uses the attn-TP group too.
- Add a per-layer `LayerCommunicator` that gathers the DP-local, attn-TP-partial output into the full-TP layout for MoE/MLP and scatters it back (`TP_ATTN_FULL` ↔ `FULL`).
- Handle DP padding rows end-to-end: trim / zero the padding so it never writes KV cache, updates Mamba recurrent state, or leaks into residuals; thread the real (unpadded) batch size and prefill-token count through the Mamba metadata and the CUDA-graph replay.
- `o_proj` / `out_proj` run with `reduce_results=False` plus chunked, global-TP-emulated fp32 accumulation so the deferred reduction matches the non-DP numerics.

## Accuracy Tests

gsm8k smoke test on a Nemotron model (tp=8, eager, `parallel=32`, 200 questions):

| config             | accuracy | invalid |
| ------------------ | -------- | ------- |
| no DP, no MTP      | 0.955    | 0.000   |
| DP attention only  | 0.970    | 0.000   |
| MTP only           | 0.955    | 0.000   |
| DP attention + MTP | 0.955    | 0.000   |

DP attention + MTP matches the no-DP/no-MTP and single-feature baselines (no accuracy regression, no hangs).

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :white_check_mark: [Run #27378689629](https://github.com/sgl-project/sglang/actions/runs/27378689629)
Latest PR Test (Extra): :white_check_mark: [Run #27437811766](https://github.com/sgl-project/sglang/actions/runs/27437811766)

### CODE REVIEW

File: python/sglang/srt/models/nemotron_h.py [hunk 1]
```diff
@@ -86,6 +98,117 @@
...
+
+def _is_sparse_layer(layer_type: str) -> bool:
+    return layer_type == MOE
+
+
+def _is_attn_layer(layer_type: str) -> bool:
+    return layer_type in ATTN_LAYERS
+
+
+def _get_real_num_tokens(
```
Comment (b8zhong) [inline python/sglang/srt/models/nemotron_h.py]: Please move these functions outside of the modeling file

File: python/sglang/srt/models/nemotron_h.py [hunk 2]
```diff
@@ -86,6 +98,117 @@
...
+        if offset + actual < hidden_states.shape[0]:
+            end = min(offset + padded, hidden_states.shape[0])
+            hidden_states[offset + actual : end].zero_()
+        offset += padded
+        if offset >= hidden_states.shape[0]:
+            break
+    return hidden_states
+
+
+def _pad_to_original_num_tokens(
```
Comment (b8zhong) [inline python/sglang/srt/models/nemotron_h.py]: There  are a lot of zero and pad functions. Is that expected?

File: python/sglang/srt/models/nemotron_h.py [hunk 3]
```diff
@@ -513,10 +715,45 @@ def __init__(
...
+            attn_output = self.attn.forward(q, k, v, forward_batch)
+            output, _ = self.o_proj(attn_output)
+            return output
+
+        padded_shape = hidden_states.shape[0]
+        real_tokens = _get_real_num_tokens(hidden_states, forward_batch)
+        original_out_cache_loc = forward_batch.out_cache_loc
+        trim_for_attention = real_tokens < padded_shape
+        trim_q_for_attention = trim_for_attention and not forward_batch.forward_mode.is_decode()
+        if trim_for_attention and original_out_cache_loc is not None:
```
Comment (b8zhong) [inline python/sglang/srt/models/nemotron_h.py]: Can you please verify the logic here, it looks quite complicated

File: python/sglang/srt/model_executor/forward_batch_info.py [hunk 4]
```diff
@@ -519,13 +519,13 @@ def init_new(
                 global_num_tokens = batch.global_num_tokens
                 global_num_tokens_for_logprob = batch.global_num_tokens_for_logprob
 
-            ret.original_global_num_tokens_cpu = batch.global_num_tokens
-            ret.global_num_tokens_cpu = global_num_tokens
+            ret.original_global_num_tokens_cpu = list(batch.global_num_tokens)
```
Comment (b8zhong) [inline python/sglang/srt/model_executor/forward_batch_info.py]: Why do we need to change these to list?

File: python/sglang/srt/model_executor/cuda_graph_runner.py [hunk 5]
```diff
@@ -1284,17 +1286,22 @@ def replay_prepare(
...
-            buffers.seq_lens[:bs],
-            forward_batch.seq_lens_sum + (bs - raw_bs) * self.seq_len_fill_value,
-            buffers.encoder_lens[:bs] if self.is_encoder_decoder else None,
-            self.capture_forward_mode,
-            forward_batch.spec_info,
-            seq_lens_cpu=buffers.seq_lens_cpu[:bs],
-        )
-        attn_backend._replay_forward_batch = None
+        setattr(attn_backend, "_replay_num_decodes", raw_bs)
+        try:
```
Comment (b8zhong) [inline python/sglang/srt/model_executor/cuda_graph_runner.py]: Can we please avoid try finally or try except in this path

File: python/sglang/srt/layers/attention/hybrid_linear_attn_backend.py [hunk 6]
```diff
@@ -803,17 +804,26 @@ def init_forward_metadata_replay_cuda_graph(
...
-                seq_lens,
-                seq_lens_sum,
-                encoder_lens,
-                forward_mode,
-                spec_info,
-                seq_lens_cpu,
-            )
+            if replay_num_decodes is not None:
+                setattr(attn_backend, "_replay_num_decodes", replay_num_decodes)
+            try:
```
Comment (b8zhong) [inline python/sglang/srt/layers/attention/hybrid_linear_attn_backend.py]: Do not use try finally

File: python/sglang/srt/layers/linear.py [hunk 7]
```diff
@@ -1535,8 +1535,41 @@ def forward(self, input_, skip_all_reduce=False, forward_batch=None):
...
+                    input_chunks[1:], weight_chunks[1:]
+                ):
+                    output_parallel = output_parallel + torch.nn.functional.linear(
+                        input_chunk,
+                        weight_chunk,
+                        None,
+                    ).float()
+                output_parallel = output_parallel.to(output_dtype)
+            else:
+                output_parallel = self.quant_method.apply(self, input_parallel, bias=bias_)
```
[the remark this answers is not in the document]
Comment (b8zhong) [reply python/sglang/srt/layers/linear.py]: Please verify if this is true (it looks like a reasonable comment)

File: python/sglang/srt/model_executor/model_runner.py [hunk 8]
```diff
@@ -3351,11 +3349,16 @@ def _forward_raw(
...
+        # Keep num_token_non_padded in the same (DP-local, pre-padding)
+        # semantics as num_token_non_padded_cpu. DP-attention model paths run
+        # attention/Mamba on TP-attn-full replicated local rows and gather only
+        # at the MLP/MoE boundary; slicing this scalar by attention TP rank makes
+        # TP rank 1 see zero real tokens for bs=1 decode under tp=8,dp=4.
         if (
             forward_batch.num_token_non_padded is not None
             and forward_batch.global_num_tokens_gpu is not None
             and require_gathered_buffer(self.server_args)
+            and not self.server_args.enable_dp_attention
```
Comment (Fridge003) [inline python/sglang/srt/model_executor/model_runner.py]: This operation is unexpected. When dp is enabled we also need to adjust token num for tp

File: python/sglang/srt/model_executor/forward_batch_info.py [hunk 9]
```diff
@@ -519,13 +519,17 @@ def init_new(
                 global_num_tokens = batch.global_num_tokens
                 global_num_tokens_for_logprob = batch.global_num_tokens_for_logprob
 
-            ret.original_global_num_tokens_cpu = batch.global_num_tokens
-            ret.global_num_tokens_cpu = global_num_tokens
+            # Copy into independent lists: original_* is the pre-DP-padding
```
Comment (Fridge003) [inline python/sglang/srt/model_executor/forward_batch_info.py]: We should avoid any modification under `model_executor` folder unless necessary. If necessary, protect it under an if branch for nemotron model

File: python/sglang/srt/layers/linear.py [hunk 10]
```diff
@@ -1535,8 +1535,50 @@ def forward(self, input_, skip_all_reduce=False, forward_batch=None):
...
                 get_tp_group(), disabled=not is_allocation_symmetric()
             )
+        # emulate_global_tp_chunks (opt-in via the per-layer attribute, e.g. the
+        # DP-attention o_proj/out_proj which run on the attn-TP group): split the
+        # local matmul into global_tp/attn_tp chunks and accumulate in fp32 so the
+        # fp reduction granularity matches a full global-TP run. This is NOT a
+        # no-op vs a single GEMM (same sum, different fp accumulation order); it
+        # exists for global-TP numerical parity (batch-invariance / train-inference
+        # consistency). Other layers never set the attribute and skip this path.
+        emulate_global_tp_chunks = 1
```
Comment (Fridge003) [inline python/sglang/srt/layers/linear.py]: Do we really need this emulate_global_tp_chunks modification

File: python/sglang/srt/models/nemotron_h_utils.py [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+    LayerScatterModes,
+    ScatterMode,
+)
+from sglang.srt.layers.layernorm import RMSNorm
+from sglang.srt.model_executor.forward_batch_info import ForwardBatch
+
+ATTN_LAYERS = (MAMBA, ATTENTION)
+
+
+def _is_sparse_layer(layer_type: str) -> bool:
```
Comment (Oasis-Git) [inline python/sglang/srt/models/nemotron_h_utils.py]: we may not need this?

File: python/sglang/srt/models/nemotron_h_utils.py
Comment (Oasis-Git) [inline python/sglang/srt/models/nemotron_h_utils.py:1]: should this file can be used for all mamba attn model

File: python/sglang/srt/models/nemotron_h.py [hunk 12]
```diff
@@ -436,6 +491,30 @@ def forward(
...
+        if is_dp_attention_enabled():
+            if self.prev_layer_is_attn and residual is not None:
+                hidden_states = attn_tp_all_reduce(hidden_states)
+            hidden_states, residual = _zero_dp_padding_rows(
+                hidden_states, forward_batch, residual
+            )
+            hidden_states, residual = self.layer_communicator.prepare_attn(
+                hidden_states, residual, forward_batch
+            )
+            if (
```
Comment (Oasis-Git) [inline python/sglang/srt/models/nemotron_h.py:486]: Why we need to add logics for idle?

File: python/sglang/srt/model_executor/forward_batch_info.py [hunk 13]
```diff
@@ -962,11 +971,10 @@ def prepare_mlp_sync_batch(self, model_runner: ModelRunner):
                 self.extend_start_loc = torch.arange(
                     bs, dtype=torch.int32, device=self.seq_lens.device
                 )
-                self.extend_prefix_lens_cpu = self.extend_prefix_lens.cpu()
-                self.extend_seq_lens_cpu = self.extend_seq_lens.cpu()
+                self.extend_prefix_lens_cpu = self.extend_prefix_lens.cpu().tolist()
```
Comment (Oasis-Git) [inline python/sglang/srt/model_executor/forward_batch_info.py]: why tolist() here is necessary

**Review by Oasis-Git — COMMENTED** [review]
Hi left some comments here

**Comment by Oasis-Git** [comment]
Ran GSM8K at temperature=0 against three configs, full 1314 examples, 8× H200, same hardware throughout. The PR diff was applied on top of sgl-project/sglang:main with no other changes.

| Config              | TP | DP | Total latency | Output throughput | GSM8K |
  | ------------------- | -- | -- | ------------- | ----------------- | ----- |
  | main, no DP         |  8 | —  |       84.1 s  |       6211 tok/s  | 0.968 |
  | PR + no DP          |  8 | —  |       89.6 s  |       5823 tok/s  | 0.968 |
  | PR + DP (mean of 3) |  8 | 4  |      118.3 s  |       4457 tok/s  | 0.968 |

  Accuracy is unchanged across all configs (steady-state mean of the DP path = 0.9678; main baseline = 0.9680 — within run-to-run NCCL non-determinism of ~1 problem out of 1314). The first DP run came in slower and at 0.961, which was
  reproducibly a cold-state artifact: subsequent runs against the same server settled at the numbers shown above.

File: python/sglang/srt/model_executor/model_runner.py [hunk 14]
```diff
@@ -3307,10 +3307,14 @@ def _forward_raw(
...
             # (Qwen3/Qwen2 MoE) keeps the attn_tp-replicated layout and wants the
             # adjustment to run — see docs/design/prefill-cp-mla.md §Phase 5.
+            # Also skip under DP attention: those paths run attention/Mamba on a
+            # TP-attn-full replicated local layout and only gather at the MLP/MoE
+            # boundary, so slicing by attn_tp_size would zero rank 1 for bs=1 decode.
             if (
                 forward_batch.num_token_non_padded is not None
                 and forward_batch.global_num_tokens_gpu is not None
                 and require_gathered_buffer(self.server_args)
+                and not self.server_args.enable_dp_attention
```
Comment (Fridge003) [inline python/sglang/srt/model_executor/model_runner.py]: Adding this condition looks risky, need @ch-wan to take a look

File: python/sglang/srt/model_executor/forward_batch_info.py (hunk 13 above)
Comment (Fridge003) [reply python/sglang/srt/model_executor/forward_batch_info.py]: +1

Comment (Zhichenzzz) [reply python/sglang/srt/model_executor/forward_batch_info.py]: solved thanks

**Comment by Fridge003** [comment]
cc @ch-wan  @ispobock  for further reviews

File: python/sglang/srt/model_executor/forward_batch_info.py [hunk 15]
```diff
@@ -1135,20 +1138,39 @@ def prepare_mlp_sync_batch(self, model_runner: ModelRunner):
             or self.forward_mode.is_draft_extend(include_v2=True)
             or self.forward_mode.is_idle()
         ):
-            if self.is_extend_in_batch and dp_padding_mode.is_max_len():
+            if self.spec_info is not None and not self.spec_info.is_draft_input():
+                if self.forward_mode.is_idle():
+                    setattr(self, "_original_forward_mode", self.forward_mode)
```
Comment (ispobock) [inline python/sglang/srt/model_executor/forward_batch_info.py]: It seems a little hacky here to use dynamic attr and silent state mutation for forward mode.

File: python/sglang/srt/models/nemotron_h.py [hunk 16]
```diff
@@ -76,6 +82,14 @@
     replace_prefix,
     replace_substrings,
 )
+from sglang.srt.models.nemotron_h_utils import (
+    _get_real_num_tokens,
```
Comment (ispobock) [inline python/sglang/srt/models/nemotron_h.py]: why import private functions?

File: python/sglang/srt/model_executor/forward_batch_info.py (hunk 15 above)
Comment (Zhichenzzz) [reply python/sglang/srt/model_executor/forward_batch_info.py]: Good catch! fixed

File: python/sglang/srt/models/nemotron_h.py (hunk 16 above)
Comment (Zhichenzzz) [reply python/sglang/srt/models/nemotron_h.py]: renamed thoes

**Review by Fridge003 — APPROVED** [review]
